src/gamestate/commands.cpp

```diff
@@ -3970,13 +3970,26 @@ void execute_toggle_mobilization(sys::state& state, dcon::nation_id source) {
 	}
 }
 
+static void post_chat_message(sys::state& state, ui::chat_message& m) {
+	// Private message
+	bool can_see = true;
+	if(bool(m.target)) {
+		can_see = state.local_player_nation == m.source || state.local_player_nation == m.target;
+	}
+	if(can_see) {
+		state.ui_state.chat_messages[state.ui_state.chat_messages_index++] = m;
+		if(state.ui_state.chat_messages_index >= state.ui_state.chat_messages.size())
+			state.ui_state.chat_messages_index = 0;
+	}
+}
+
 void chat_message(sys::state& state, dcon::nation_id source, std::string_view body, dcon::nation_id target) {
 	payload p;
 	memset(&p, 0, sizeof(payload));
 	p.type = command_type::chat_message;
 	p.source = source;
 	p.data.chat_message.target = target;
-	memcpy(p.data.chat_message.body, std::string(body).c_str(), ui::max_chat_message_len);
+	memcpy(p.data.chat_message.body, std::string(body).c_str(), std::min<size_t>(body.length() + 1, size_t(ui::max_chat_message_len)));
```

Comment:
You don't need to zero terminate all the strings if you want to fit an extra char in there